workflow-core/src/main/java/com/squareup/workflow/CoroutineWorkflow.kt

```diff
@@ -0,0 +1,82 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.squareup.workflow
+
+import kotlinx.coroutines.experimental.CoroutineScope
+import kotlinx.coroutines.experimental.Deferred
+import kotlinx.coroutines.experimental.async
+import kotlinx.coroutines.experimental.channels.BroadcastChannel
+import kotlinx.coroutines.experimental.channels.Channel
+import kotlinx.coroutines.experimental.channels.Channel.Factory.CONFLATED
+import kotlinx.coroutines.experimental.channels.Channel.Factory.UNLIMITED
+import kotlinx.coroutines.experimental.channels.ClosedSendChannelException
+import kotlinx.coroutines.experimental.channels.ReceiveChannel
+import kotlinx.coroutines.experimental.channels.SendChannel
+import kotlin.coroutines.experimental.CoroutineContext
+import kotlin.coroutines.experimental.EmptyCoroutineContext
+
+/**
+ * Starts a coroutine in the given [context] that runs [block] to produce states and the result
+ * for the returned [Workflow].
+ *
+ * The block gets a [SendChannel] that it can use to emit [states][S], and a [ReceiveChannel] that
+ * it can use to receive [events][E] from the workflow's [WorkflowInput]. The block's return value
+ * is used as the workflow's result.
+ * The state channel is a
+ * [ConflatedBroadcastChannel][kotlinx.coroutines.experimental.channels.ConflatedBroadcastChannel],
+ * which means that sends will never suspend or fail, and if observers of the workflow aren't
+ * receiving fast enough, they will miss intermediate states.
+ * The event channel has [unlimited capacity][UNLIMITED].
+ */
+fun <S : Any, E : Any, O : Any> CoroutineScope.workflow(
+  context: CoroutineContext = EmptyCoroutineContext,
+  block: suspend CoroutineScope.(
+    state: SendChannel<S>,
+    events: ReceiveChannel<E>
+  ) -> O
+): Workflow<S, E, O> {
+  // The events channel is buffered primarily so that reactors don't race themselves when state
+  // changes immediately cause more events to be sent.
+  val events = Channel<E>(UNLIMITED)
+  val state = BroadcastChannel<S>(CONFLATED)
+  val result = async(context) {
+    block(state, events)
+  }
+
+  result.invokeOnCompletion { cancelReason ->
+    state.close(cancelReason)
+    // Don't hold onto any remaining queued events, since they will never be read.
+    events.close(cancelReason)
+  }
+
+  return object : Workflow<S, E, O>, Deferred<O> by result {
+    override fun openSubscriptionToState(): ReceiveChannel<S> = state.openSubscription()
+
+    override fun sendEvent(event: E) {
+      // The events channel has an unlimited-size buffer, this will only fail
+      // if the event channel was closed.
+      try {
+        events.offer(event)
+      } catch (e: ClosedSendChannelException) {
+        // This may mean the workflow was canceled or finished, or that the workflow closed the
```

Comment:
I wonder how often we'll find ourselves crashing due to late events being aimed at a finished workflow. Hopefully those will be actual bugs, not something we'll need to become more forgiving about. (Not suggesting any changes, just chatting.)